### PR TITLE
Add a pre-build shell script step to all Maven and Frestyle Projects

### PR DESCRIPTION
**Pull request by kuisathaverat**
Example script that add pre build shell script step to all Maven and Freestyle Projects

@reviewbybees

### CODE REVIEW

File: scriptler/addShellPreBuildStepToProject.groovy [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+	.each{
+      println it.name + " - " + it.class
+
+      DescribableList<Builder,Descriptor<Builder>> builders = new DescribableList<Builder,Descriptor<Builder>>()
+      builders.add(new hudson.tasks.Shell("echo 'Hello world'"))
+
+      if(it instanceof FreeStyleProject){
+        builders.addAll(it.getBuildersList())
+        it.getBuildersList().clear()
+        it.getBuildersList()addAll(builders)
```
Comment (Dohbedoh) [inline scriptler/addShellPreBuildStepToProject.groovy]: There is a `.` missing here. `it.getBuildersList()addAll(builders)` should be `it.getBuildersList().addAll(builders)`

File: scriptler/addShellPreBuildStepToProject.groovy [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+import hudson.maven.*
+
+
+Jenkins.instance.getAllItems(Job.class)
+	.findAll{ it instanceof FreeStyleProject || it instanceof MavenModuleSet  }
+	.each{
+      println it.name + " - " + it.class
+
+      DescribableList<Builder,Descriptor<Builder>> builders = new DescribableList<Builder,Descriptor<Builder>>()
+      builders.add(new hudson.tasks.Shell("echo 'Hello world'"))
```
Comment (abayer) [inline scriptler/addShellPreBuildStepToProject.groovy]: :bug: Given that this is pretty clearly just an example, I'm not sure we want it in the Scriptler catalog as is - if you want to make it just do shell steps and have the shell script as a parameter, that'd work.

File: scriptler/addShellPreBuildStepToProject.groovy [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+import hudson.util.*
+import hudson.tasks.*
+import hudson.maven.*
+
+//def filePath = '/path/to/script.sh'
+def scriptFile = new File(filePath)
+
+if(!scriptFile.exists() && scriptFile.isDirectory()) { 
+  throw IOException("Check file name")
```
Comment (Evildethow) [inline scriptler/addShellPreBuildStepToProject.groovy]: Syntax issue

`throw IOException("Check file name")` should be `throw new IOException("Check file name")`

File: scriptler/addShellPreBuildStepToProject.groovy [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+}
+String script = scriptFile.text
+
+Jenkins.instance.getAllItems(Job.class)
+	.findAll{ it instanceof FreeStyleProject || it instanceof MavenModuleSet  }
+	.each{
+      println it.name + " - " + it.class
+
+      DescribableList<Builder,Descriptor<Builder>> builders = new DescribableList<Builder,Descriptor<Builder>>()
```
Comment (Evildethow) [inline scriptler/addShellPreBuildStepToProject.groovy]: Im surprised this works as the constructor is `protected` scope. 

```
    protected DescribableList() {
    }
```

But assuming it does so just a comment.

Comment (kuisathaverat) [inline scriptler/addShellPreBuildStepToProject.groovy]: Groovy Magic

File: scriptler/addShellPreBuildStepToProject.groovy (hunk 3 above)
Comment (kuisathaverat) [inline scriptler/addShellPreBuildStepToProject.groovy]: the groovy interpreter eats everything

**Comment by kuisathaverat** [comment]
@Evildethow @Dohbedoh @abayer Could it be good now?

**Comment by aheritier** [comment]
LGTM

**Pull request merged**